WIP: Removed format specific Timesteps

richardjgowers opened the pull request:
Fixes #1575

So this is squishing all the format specific Timesteps we have.  This mostly involves moving the conversion of unitcell information from the Timestep to the Reader.

Theres's only one custom Timestep remaining for MemoryReader.  This works slightly differently (I think?) because `Timestep.positions` provides a view onto the big chunk of memory that the MemoryReader operates on.

Very much work in progress... @kain88-de thoughts?

kain88-de commented on package/MDAnalysis/coordinates/DCD.py:
```diff
@@ -146,7 +146,7 @@ def __init__(self, filename, convert_units=True, dt=None, **kwargs):
 
         self._ts_kwargs['dt'] = dt
-        self.ts = self._Timestep(self.n_atoms, **self._ts_kwargs)
+        self.ts = base.Timestep(self.n_atoms, **self._ts_kwargs)
```
I didn't delete that already?

kain88-de replied:
Shouldn't we also delete `self._Timestep` then?

richardjgowers replied:
yeah self._Timestep is a thing of the past

kain88-de commented on package/MDAnalysis/coordinates/GRO.py:
```diff
@@ -163,23 +140,23 @@ def _read_first_frame(self):
                 # 2 header lines, 1 box line at end
                 if pos == n_atoms:
-                    unitcell = np.float32(line.split())
+                    unitcell = list(map(float, line.split()))
```
I thought we benchmarked those to use `np.float32`?

orbeckst replied:
Yes, @kain88-de is right. We determined that the fastest conversion is letting numpy do everything.

orbeckst replied:
yes

orbeckst commented on package/MDAnalysis/coordinates/DLPoly.py:
```diff
@@ -125,17 +110,17 @@ def _read_first_frame(self):
             ts._forces = forces
         if not imcon == 0:
-            ts._unitcell = unitcell
+            ts.triclinic_dimensions = unitcell
```
Why add `Timestep.triclinic_dimensions` ? Is this the "original" unitcell – and if so, why not keep it as a private `_unitcell`. The user should not need to worry about box representations. We simply provide `ts.dimensions` and that's it.

orbeckst commented on package/MDAnalysis/coordinates/DLPoly.py:
```diff
@@ -159,23 +143,24 @@ def __init__(self, filename, **kwargs):
+            unitcell[0] = self._file.readline().split()
+            unitcell[1] = self._file.readline().split()
+            unitcell[2] = self._file.readline().split()
+            ts.triclinic_dimensions = unitcell
```
I don't see the point of `ts.triclinic_dimensions`

orbeckst commented on package/MDAnalysis/coordinates/GRO.py:
```diff
@@ -97,47 +94,28 @@
-
-    def _init_unitcell(self):
-        return np.zeros(9, dtype=np.float32)
+"""unitcell dimensions (A, B, C, alpha, beta, gamma)
```
Should these docs be properly formatted for module level docs? From the diff I cannot really see what's going on.

richardjgowers replied:
yeah this is an old comment from withing the format specific timestep, I kept it around as it seemd useful for developers

orbeckst commented on package/MDAnalysis/coordinates/PDBQT.py:
```diff
@@ -163,9 +163,7 @@ def _read_first_frame(self):
-            coords,
-            **self._ts_kwargs)
+        self.ts = base.Timestep.from_coordinates(coords, **self._ts_kwargs)
         self.ts._unitcell[:] = unitcell
```
Didn't you want to remove `_unitcell`?

orbeckst commented on package/MDAnalysis/coordinates/PQR.py:
```diff
@@ -132,8 +132,7 @@ def _read_first_frame(self):
-        self.ts = self._Timestep.from_coordinates(
-            coords, **self._ts_kwargs)
+        self.ts = base.Timestep.from_coordinates(coords, **self._ts_kwargs)
         self.ts._unitcell[:] = unitcell
```
Didn't you want to remove `_unitcell`?

orbeckst commented on package/MDAnalysis/coordinates/TRZ.py:
```diff
@@ -37,45 +37,6 @@
-.. autoclass:: MDAnalysis.coordinates.TRZ.Timestep
-   :members:
-
-   .. attribute:: frame
```
are all of these special values now in `Timestep.data`?

orbeckst commented on package/MDAnalysis/coordinates/base.py:
```diff
@@ -274,7 +273,9 @@ def __init__(self, n_atoms, **kwargs):
-        self._unitcell = self._init_unitcell()
+        self.order = kwargs.get('order', 'F')
+
+        self._unitcell = np.zeros(6, dtype=np.float32)
```
What does `_unitcell` hold now if it's not the original box?

kain88-de commented:
[gist](https://gist.github.com/kain88-de/c32bcc6ae70610ae0b1cf1631b2e21f1)

I did a quick check. If we use double precision for unitcell representation then all conversions between triclinic and box are accurate up to 8 decimal places. This should be enough for us and mean we can skip keeping track of the original since the two are very very close.

richardjgowers commented:
Ok so to clarify this PR, 

**Input**: `ts._unitcell` is the internal storage of the unit cell.  It is currently is the [a, b, c, alpha, beta, gamma] representation for ALL Readers, (because we've killed off custom Timestep objects).  This means that ALL Readers now have to fill `_unitcell` with information in that form.

**Access** `ts.dimensions` returns the same [a, b, c, alpha, beta, gamma] representation. `ts.triclinic_dimensions` returns the 3x3 array version of the box instead.  By having these as properties we can allow settings (eg `ts.triclinic_dimensions = ` will populate `ts._unitcell` properly.

But having reread #1575, if we want `ts.dimensions` to return a view/pointer, we can't really keep `ts.triclinic_dimensions`, as it won't return a view/pointer and will lead to confusion later on.  So I think I'm going to have to deprecate `triclinic_dimensions`.

@kain88-de I'm guessing that single precision doesn't work well enough?  So dimensions should always be float64?

kain88-de commented:
> @kain88-de I'm guessing that single precision doesn't work well enough? So dimensions should always be float64?

yeah. You can replace the functions at the end of the notebook with the equivalents of the last stable release (add `mda.lib.mdamath.` at the beginning). Then the tests fail. The only change I made in comparison to the code in MDAnalysis was to remove the type definitions to single precision. 



> But having reread #1575, if we want ts.dimensions to return a view/pointer, we can't really keep ts.triclinic_dimensions, as it won't return a view/pointer and will lead to confusion later on. So I think I'm going to have to deprecate triclinic_dimensions.

We don't we allow `ts.dimensions` and `ts.tricilinic_vectors` to have a setters that stores the result in a `ts._dim` variable that gets updated by a copy with `self._dim[:] = dims`. Then we can keep the functions and we only do a copy on the assigment and reading is still a faster reference. 

It doesn't solve #1575 for which the only solution is to generate a new timestep on a read instead of updating an old one.

richardjgowers commented:
@kain88-de I think what you call `ts._dim` is just what `ts._unitcell` is currently, both `ts.dimensions` and `ts.triclinic_dimensions` have setters which update `ts._unitcell`

kain88-de commented:
@richardjgowers do we want to forbid setting `_unitcell` directly? We could add a setter for it that throws an exception when it's called. That way we are consistent with the usage within the library?

richardjgowers commented:
@kain88-de it's a "private" attribute, if you're setting it directly then you're opening the door to weird behaviour and tears and that's not our problem :) We should probably rewrite the Readers to behave WRT that though

kain88-de commented:
Yeah true. But it might help to find which reader is a bad boy that manipulates the private attribute directly. 

> On 1. Feb 2018, at 14:55, Richard Gowers <notifications@github.com> wrote:
> 
> @kain88-de it's a "private" attribute, if you're setting it directly then you're opening the door to weird behaviour and tears and that's not our problem :) We should probably rewrite the Readers to behave WRT that though
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

richardjgowers commented:
Yeah I think I need to go through and make sure all unitcell info is double

On Fri, 2 Feb 2018, 12:46 Max Linke, <notifications@github.com> wrote:

> *@kain88-de* commented on this pull request.
> ------------------------------
>
> In package/MDAnalysis/coordinates/GRO.py
> <https://github.com/MDAnalysis/mdanalysis/pull/1765#discussion_r165635583>
> :
>
> > @@ -147,7 +147,7 @@ def _read_first_frame(self):
>              for pos, line in enumerate(grofile, start=-2):
>                  # 2 header lines, 1 box line at end
>                  if pos == n_atoms:
> -                    unitcell = list(map(float, line.split()))
> +                    unitcell = np.float32(line.split())
>
> lets directly use double precision.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/MDAnalysis/mdanalysis/pull/1765#pullrequestreview-93609482>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AI0jB1saMJuQuexyhUscrNSvpYk6DR6Hks5tQwO4gaJpZM4RyoNU>
> .
>

richardjgowers commented:
tangled mess of trying to reduce classes...
